Only in b: .config
diff -r -u a/drivers/hid/usbhid/hid-quirks.c b/drivers/hid/usbhid/hid-quirks.c
--- a/drivers/hid/usbhid/hid-quirks.c	2008-07-14 01:51:29.000000000 +0400
+++ b/drivers/hid/usbhid/hid-quirks.c	2009-01-22 00:13:28.000000000 +0300
@@ -390,6 +390,9 @@
 #define USB_VENDOR_ID_TURBOX		0x062a
 #define USB_DEVICE_ID_TURBOX_KEYBOARD	0x0201
 
+#define USB_VENDOR_ID_UC_LOGIC		0x5543
+#define USB_DEVICE_ID_UC_LOGIC_WP8060U	0x0005
+
 #define USB_VENDOR_ID_VERNIER		0x08f7
 #define USB_DEVICE_ID_VERNIER_LABPRO	0x0001
 #define USB_DEVICE_ID_VERNIER_GOTEMP	0x0002
@@ -575,6 +578,7 @@
 	{ USB_VENDOR_ID_PANJIT, 0x0003, HID_QUIRK_IGNORE },
 	{ USB_VENDOR_ID_PANJIT, 0x0004, HID_QUIRK_IGNORE },
 	{ USB_VENDOR_ID_SOUNDGRAPH, USB_DEVICE_ID_SOUNDGRAPH_IMON_LCD, HID_QUIRK_IGNORE },
+ 	{ USB_VENDOR_ID_UC_LOGIC, USB_DEVICE_ID_UC_LOGIC_WP8060U, HID_QUIRK_MULTI_INPUT },
 	{ USB_VENDOR_ID_VERNIER, USB_DEVICE_ID_VERNIER_LABPRO, HID_QUIRK_IGNORE },
 	{ USB_VENDOR_ID_VERNIER, USB_DEVICE_ID_VERNIER_GOTEMP, HID_QUIRK_IGNORE },
 	{ USB_VENDOR_ID_VERNIER, USB_DEVICE_ID_VERNIER_SKIP, HID_QUIRK_IGNORE },
@@ -744,6 +748,8 @@
 	{ USB_VENDOR_ID_CYPRESS, USB_DEVICE_ID_CYPRESS_BARCODE_1, HID_QUIRK_RDESC_SWAPPED_MIN_MAX },
 	{ USB_VENDOR_ID_CYPRESS, USB_DEVICE_ID_CYPRESS_BARCODE_2, HID_QUIRK_RDESC_SWAPPED_MIN_MAX },
 
+	{ USB_VENDOR_ID_UC_LOGIC, USB_DEVICE_ID_UC_LOGIC_WP8060U, HID_QUIRK_RDESC_UC_LOGIC_WP8060U },
+
 	{ 0, 0 }
 };
 
@@ -1135,6 +1141,30 @@
 	}
 }
 
+/*
++ * UC-LOGIC WP8060U HID report descriptor has an extra input which is
++ * invalid, never used and will result in an extra device when combined with
++ * HID_QUIRK_MULTI_INPUT.
++ */
+static void usbhid_fixup_uc_logic_descriptor(char *rdesc, int rsize)
+{
+	int new_rsize;
+
+	/* Minimal sanity check */
+	if (rsize != 212)
+		return;
+
+	printk(KERN_INFO "Fixing up UC-LOGIC tablet report descriptor\n");
+
+	/*
+	 * Cut off the first input descriptor (out of 3)
+	 * which is incorrect and is never used.
+	 */
+	new_rsize = rsize - 89;
+	memmove(rdesc, rdesc + (rsize - new_rsize), new_rsize);
+	memset(rdesc + new_rsize, 0, (rsize - new_rsize));
+}
+
 static void __usbhid_fixup_report_descriptor(__u32 quirks, char *rdesc, unsigned rsize)
 {
 	if ((quirks & HID_QUIRK_RDESC_CYMOTION))
@@ -1163,6 +1193,9 @@
 
 	if (quirks & HID_QUIRK_RDESC_SUNPLUS_WDESKTOP)
 		usbhid_fixup_sunplus_wdesktop(rdesc, rsize);
+
+	if (quirks & HID_QUIRK_RDESC_UC_LOGIC_WP8060U)
+		 usbhid_fixup_uc_logic_descriptor(rdesc, rsize);
 }
 
 /**
diff -r -u a/include/linux/hid.h b/include/linux/hid.h
--- a/include/linux/hid.h	2008-07-14 01:51:29.000000000 +0400
+++ b/include/linux/hid.h	2009-01-21 23:40:56.000000000 +0300
@@ -298,6 +298,7 @@
 #define HID_QUIRK_RDESC_SAMSUNG_REMOTE		0x00000040
 #define HID_QUIRK_RDESC_MICROSOFT_RECV_1028	0x00000080
 #define HID_QUIRK_RDESC_SUNPLUS_WDESKTOP	0x00000100
+#define HID_QUIRK_RDESC_UC_LOGIC_WP8060U	0x00000200
 
 /*
  * This is the global environment of the parser. This information is
